### PR TITLE
Treeshaking + LazyLoading for video effects package

### PR DESCRIPTION
**Pull request by prabhjot-msft**
# What
Treeshaking + LazyLoading for video effects package

# Why
https://skype.visualstudio.com/SPOOL/_workitems/edit/3336304

# How Tested

# Process & policy checklist

- [ ] I have updated the project documentation to reflect my changes if necessary.
- [ ] I have read the [CONTRIBUTING](https://github.com/Azure/communication-ui-library/blob/main/CONTRIBUTING.md) documentation.

**Is this a breaking change?**

- [ ] This change causes current functionality to break.

### CODE REVIEW

File: packages/communication-react/review/beta/communication-react.api.md [hunk 1]
```diff
@@ -3112,6 +3123,12 @@ export type OnRenderAvatarCallback = (
 userId?: string, options?: CustomAvatarOptions,
 defaultOnRender?: (props: CustomAvatarOptions) => JSX.Element) => JSX.Element;
 
+// @public (undocumented)
+export const onResolveVideoEffectDependency: () => Promise<VideoBackGroundDependency>;
+
+// @public (undocumented)
```
Comment (PorterNan) [inline packages/communication-react/review/beta/communication-react.api.md:3129]: Please add comment for all those undocumented api

File: packages/communication-react/review/beta/communication-react.api.md [hunk 2]
```diff
@@ -3112,6 +3123,12 @@ export type OnRenderAvatarCallback = (
 userId?: string, options?: CustomAvatarOptions,
 defaultOnRender?: (props: CustomAvatarOptions) => JSX.Element) => JSX.Element;
 
+// @public (undocumented)
```
Comment (PorterNan) [inline packages/communication-react/review/beta/communication-react.api.md:3126]: Do we expose lazy loading as public or beta? I see some of them marked as public but other marked as beta

File: packages/communication-react/review/stable/communication-react.api.md [hunk 3]
```diff
@@ -1423,7 +1427,7 @@ export const createAzureCommunicationChatAdapter: ({ endpoint: endpointUrl, user
 export function createAzureCommunicationChatAdapterFromClient(chatClient: StatefulChatClient, chatThreadClient: ChatThreadClient): Promise<ChatAdapter>;
 
 // @public
-export const createDefaultCallingHandlers: (callClient: StatefulCallClient, callAgent: CallAgent | undefined, deviceManager: StatefulDeviceManager | undefined, call: Call | undefined) => CallingHandlers;
+export const createDefaultCallingHandlers: (callClient: StatefulCallClient, callAgent: CallAgent | undefined, deviceManager: StatefulDeviceManager | undefined, call: Call | undefined, options?: {} | undefined) => CallingHandlers;
```
Comment (PorterNan) [inline packages/communication-react/review/stable/communication-react.api.md:1430]: We should remove the options parameter if nothing is not required in stable

File: packages/communication-react/review/stable/communication-react.api.md [hunk 4]
```diff
@@ -2569,6 +2573,12 @@ export const useSelector: <ParamT extends Selector | undefined>(selector: ParamT
 // @public
 export const useTheme: () => Theme;
 
+// @beta (undocumented)
```
Comment (PorterNan) [inline packages/communication-react/review/stable/communication-react.api.md:2576]: Looks like there is an api leak in stable api.md

File: packages/react-composites/review/beta/react-composites.api.md [hunk 5]
```diff
@@ -1483,6 +1490,12 @@ export type NetworkDiagnosticChangedEvent = NetworkDiagnosticChangedEventArgs &
 // @beta
 export type OnFetchProfileCallback = (userId: string, defaultProfile?: Profile) => Promise<Profile | undefined>;
 
+// @public (undocumented)
```
Comment (PorterNan) [inline packages/react-composites/review/beta/react-composites.api.md:1493]: beta not public

File: packages/react-composites/review/beta/react-composites.api.md [hunk 6]
```diff
@@ -1483,6 +1490,12 @@ export type NetworkDiagnosticChangedEvent = NetworkDiagnosticChangedEventArgs &
...
+export const onResolveVideoEffectDependency: () => Promise<VideoBackGroundDependency>;
+
+// @public (undocumented)
```
Comment (PorterNan) [inline packages/react-composites/review/beta/react-composites.api.md:1496]: Add comment to all those undocumented apis

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts [hunk 7]
```diff
@@ -137,7 +146,10 @@ class CallContext {
       /* @conditional-compile-remove(unsupported-browser) */ environmentInfo: clientState.environmentInfo,
       /* @conditional-compile-remove(unsupported-browser) */ unsupportedBrowserVersionsAllowed: false,
       /* @conditional-compile-remove(rooms) */ roleHint: options?.roleHint,
-      /* @conditional-compile-remove(video-background-effects) */ videoBackgroundImages: options?.videoBackgroundImages,
+      /* @conditional-compile-remove(video-background-effects) */ videoBackgroundImages:
+        options?.videoBackgroundOptions?.videoBackgroundImages,
+      /* @conditional-compile-remove(video-background-effects) */
```
Comment (PorterNan) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:151]: I am thinking we should add another feature flag as "in-progress" feature other than "video-background-effects", so we can control lazy loading shipping timeframe into beta release independently

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts [hunk 8]
```diff
@@ -332,7 +346,7 @@ export class AzureCommunicationCallAdapter<AgentType extends CallAgent | BetaTea
     locator: CallAdapterLocator,
     callAgent: AgentType,
     deviceManager: StatefulDeviceManager,
-    options?: AzureCommunicationCallAdapterOptions & TeamsAdapterOptions
+    options?: AzureCommunicationCallAdapterOptions | TeamsAdapterOptions
```
Comment (PorterNan) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:349]: Thanks for catching this, '|' makes more sense for options

File: packages/react-composites/src/composites/common/resolveVideoEffectDependency.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    return new BackgroundReplacementEffect(config);
+  };
+  // eslint-disable-next-line jsdoc/require-jsdoc
+  const videoBackGroundDependency: VideoBackGroundDependency = {
+    createBackgroundBlurEffect,
+    createBackgroundReplacementEffect
+  };
+  return videoBackGroundDependency;
+};
+
```
Comment (PorterNan) [inline packages/react-composites/src/composites/common/resolveVideoEffectDependency.ts:35]: Remove this export

File: packages/react-composites/src/composites/common/resolveVideoEffectDependencyLazy.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  };
+  // eslint-disable-next-line jsdoc/require-jsdoc
+  const videoBackGroundDependency: VideoBackGroundDependency = {
+    createBackgroundBlurEffect,
+    createBackgroundReplacementEffect
+  };
+  return videoBackGroundDependency;
+};
+
+export {};
```
Comment (PorterNan) [inline packages/react-composites/src/composites/common/resolveVideoEffectDependencyLazy.ts:39]: remove the import

**Review by PorterNan — COMMENTED** [review]
Please add snapshot, add hermetic test for video effect, not required in this PR though

File: packages/calling-component-bindings/src/handlers/createCommonHandlers.ts [hunk 11]
```diff
@@ -464,7 +472,10 @@ export const createDefaultCommonCallingHandlers = memoizeOne(
         call?.localVideoStreams.find((stream) => stream.mediaStreamType === 'Video') ||
         deviceManager?.getUnparentedVideoStreams().find((stream) => stream.mediaStreamType === 'Video');
       if (stream) {
-        return stream.feature(Features.VideoEffects).startEffects(new BackgroundBlurEffect(backgroundBlurConfig));
+        const createEffect =
```
Comment (PorterNan) [inline packages/calling-component-bindings/src/handlers/createCommonHandlers.ts:475]: If someone call this handler function onBlurVideoBackground, and if there is no onResolveVideoBackGroundDependency , we should throw runtime errors

File: packages/communication-react/review/stable/communication-react.api.md (hunk 3 above)
Comment (PorterNan) [inline packages/communication-react/review/stable/communication-react.api.md:1430]: cc this options? parameter

File: samples/CallWithChat/src/app/views/CallScreen.tsx [hunk 12]
```diff
@@ -78,7 +79,12 @@ export const CallScreen = (props: CallScreenProps): JSX.Element => {
         tooltipText: 'Custom Background'
       }
     ];
-    return { videoBackgroundImages: videoBackgroundImages };
+    return {
+      videoBackgroundOptions: {
+        videoBackgroundImages,
+        onResolveDependency: onResolveVideoEffectDependencyLazy
```
Comment (PorterNan) [inline samples/CallWithChat/src/app/views/CallScreen.tsx:85]: I remember Dom proposed some names in the discussion, please check chat history about this

File: packages/calling-component-bindings/review/stable/calling-component-bindings.api.md [hunk 13]
```diff
@@ -138,7 +142,7 @@ export interface CommonCallingHandlers {
 }
 
 // @public
-export const createDefaultCallingHandlers: (callClient: StatefulCallClient, callAgent: CallAgent | undefined, deviceManager: StatefulDeviceManager | undefined, call: Call | undefined) => CallingHandlers;
+export const createDefaultCallingHandlers: (callClient: StatefulCallClient, callAgent: CallAgent | undefined, deviceManager: StatefulDeviceManager | undefined, call: Call | undefined, options?: {} | undefined) => CallingHandlers;
```
Comment (JamesBurnside) [inline packages/calling-component-bindings/review/stable/calling-component-bindings.api.md:145]: this shouldn't show in stable API

File: packages/calling-component-bindings/review/stable/calling-component-bindings.api.md [hunk 14]
```diff
@@ -233,6 +237,12 @@ export const useCallingSelector: <SelectorT extends (state: CallClientState, pro
 // @public
 export const useDeviceManager: () => StatefulDeviceManager | undefined;
 
+// @beta (undocumented)
+export type VideoEffectBackGroundDependency = {
```
Comment (JamesBurnside) [inline packages/calling-component-bindings/review/stable/calling-component-bindings.api.md:241]: BackGround -> Background (throughout pr)

File: packages/calling-component-bindings/review/stable/calling-component-bindings.api.md [hunk 15]
```diff
@@ -233,6 +237,12 @@ export const useCallingSelector: <SelectorT extends (state: CallClientState, pro
 // @public
 export const useDeviceManager: () => StatefulDeviceManager | undefined;
 
+// @beta (undocumented)
```
Comment (JamesBurnside) [inline packages/calling-component-bindings/review/stable/calling-component-bindings.api.md:240]: beta interface shouldn't show in stable api

File: change-beta/@azure-communication-react-ee550c94-de09-4920-b56c-9f5b6e1a843a.json [hunk 16]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
```
Comment (JamesBurnside) [inline change-beta/@azure-communication-react-ee550c94-de09-4920-b56c-9f5b6e1a843a.json:2]: this is a prerelease change? (for beta consumption only?)

File: change-beta/@azure-communication-react-ee550c94-de09-4920-b56c-9f5b6e1a843a.json [hunk 17]
```diff
@@ -0,0 +1,7 @@
...
+  "comment": "Changes for lazyloading + treeshaking",
```
Comment (JamesBurnside) [inline change-beta/@azure-communication-react-ee550c94-de09-4920-b56c-9f5b6e1a843a.json:3]: This comment isn't descriptive enough for a changelog

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts (hunk 7 above)
Comment (JamesBurnside) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:151]: They have to ship at the same time though no? (Unless a new feature creeps in that wants lazy-loading sooner but that feels unlikely)

File: samples/CallWithChat/src/app/views/CallScreen.tsx (hunk 12 above)
Comment (JamesBurnside) [inline samples/CallWithChat/src/app/views/CallScreen.tsx:85]: We'll need UI tests with and without this

Comment (JamesBurnside) [reply samples/CallWithChat/src/app/views/CallScreen.tsx:85]: I don't see a UI test update that no longer shows the effects button in the call screen camera button. Are we missing that test or does the effects button incorrectly show?

File: packages/calling-component-bindings/review/beta/calling-component-bindings.api.md [hunk 18]
```diff
@@ -199,10 +201,14 @@ export interface CommonCallingHandlers {
 }
 
 // @public
-export const createDefaultCallingHandlers: (callClient: StatefulCallClient, callAgent: CallAgent | undefined, deviceManager: StatefulDeviceManager | undefined, call: Call | undefined) => CallingHandlers;
+export const createDefaultCallingHandlers: (callClient: StatefulCallClient, callAgent: CallAgent | undefined, deviceManager: StatefulDeviceManager | undefined, call: Call | undefined, options?: {
+    onResolveVideoBackGroundDependency?: (() => Promise<VideoEffectBackGroundDependency>) | undefined;
```
[the remark this answers is not in the document]
Comment (prabhjot-msft) [reply packages/calling-component-bindings/review/beta/calling-component-bindings.api.md]: done

File: packages/calling-component-bindings/review/stable/calling-component-bindings.api.md (hunk 15 above)
Comment (prabhjot-msft) [reply packages/calling-component-bindings/review/stable/calling-component-bindings.api.md]: done

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts (hunk 7 above)
Comment (prabhjot-msft) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:151]: Yeah, we are shipping the dependency isolation along with video effects, so I think we can keep these under same CC

File: packages/calling-component-bindings/src/handlers/createCommonHandlers.ts [hunk 19]
```diff
@@ -121,7 +128,10 @@ export const createDefaultCommonCallingHandlers = memoizeOne(
   (
     callClient: StatefulCallClient,
     deviceManager: StatefulDeviceManager | undefined,
-    call: Call | /* @conditional-compile-remove(teams-identity-support) */ TeamsCall | undefined
+    call: Call | /* @conditional-compile-remove(teams-identity-support) */ TeamsCall | undefined,
+    options?: {
+      onResolveVideoBackgroundDependency?: () => Promise<VideoEffectBackgroundDependency>;
```
Comment (JamesBurnside) [inline packages/calling-component-bindings/src/handlers/createCommonHandlers.ts:133]: nit: inconsistent here between `VideoBackgroundDependency` and `VideoEffectBackgroundDependency`

I assume all should be `video effects background dependency` (in relevant camel/Pascal Case, note the plural effects not effect)

**Review by JamesBurnside — APPROVED** [review]

**Review by PorterNan — APPROVED** [review]
This one looks great! There are several follow-up item we can continue doing:
1. Document how to use lazy loading + webpack config/rollup config
2. Use react lazy loading method to improve the UX a bit more
3. Update snapshot - for use cases with/without video effects

**Pull request merged**